src/backend/graphics/egl/wayland.rs

```diff
@@ -0,0 +1,322 @@
+            Format::RGB | Format::RGBA | Format::External => 1,
+            Format::Y_UV | Format::Y_XUXV => 2,
+            Format::Y_U_V => 3,
+        }
+    }
+}
+
+pub struct EGLImages {
+    display: Weak<ffi::egl::types::EGLDisplay>,
+    pub width: u32,
+    pub height: u32,
+    pub y_inverted: bool,
+    pub format: Format,
+    images: Vec<EGLImage>,
+    buffer: WlBuffer,
+}
+
+impl EGLImages {
+    pub fn num_planes(&self) -> usize {
+        self.format.num_planes()
+    }
+
+    pub unsafe fn bind_to_texture(&self, plane: usize, tex_id: c_uint) -> ::std::result::Result<(), TextureCreationError> {
+        if self.display.upgrade().is_some() {
+            let mut old_tex_id: i32 = 0;
+            ffi::gl::GetIntegerv(ffi::gl::TEXTURE_BINDING_2D, &mut old_tex_id);
+            ffi::gl::BindTexture(ffi::gl::TEXTURE_2D, tex_id);
+            ffi::gl::EGLImageTargetTexture2DOES(ffi::gl::TEXTURE_2D, *self.images.get(plane).ok_or(TextureCreationError::PlaneIndexOutOfBounds)?);
+            let res = match ffi::egl::GetError() as u32 {
+                ffi::egl::SUCCESS => Ok(()),
+                err => Err(TextureCreationError::TextureBindingFailed(err)),
+            };
+            ffi::gl::BindTexture(ffi::gl::TEXTURE_2D, old_tex_id as u32);
+            res
+        } else {
+            Err(TextureCreationError::ContextLost)
+        }
+    }
+}
+
+impl Drop for EGLImages {
+    fn drop(&mut self) {
+        if let Some(display) = self.display.upgrade() {
+            for image in self.images.drain(..) {
+                unsafe { ffi::egl::DestroyImageKHR(*display, image); }
+            }
+        }
+        self.buffer.release();
+    }
+}
+
+pub trait EGLWaylandExtensions {
+    /// Binds this EGL context to the given Wayland display.
+    ///
+    /// This will allow clients to utilize EGL to create hardware-accelerated
+    /// surfaces. The server will need to be able to handle egl-wl_buffers.
+    /// See the `wayland::drm` module.
+    ///
+    /// ## Errors
+    ///
+    /// This might return `WlExtensionNotSupported` if binding is not supported
+    /// by the EGL implementation.
+    ///
+    /// This might return `OtherEGLDisplayAlreadyBound` if called for the same
+    /// `Display` multiple times, as only one context may be bound at any given time.
+    fn bind_wl_display(&self, display: &Display) -> Result<EGLDisplay>;
+
+    /// Unbinds this EGL context from the given Wayland display.
+    ///
+    /// This will stop clients from using previously available extensions
+    /// to utilize hardware-accelerated surface via EGL.
+    ///
+    /// ## Errors
+    ///
+    /// This might return `WlExtensionNotSupported` if binding is not supported
+    /// by the EGL implementation.
+    ///
+    /// This might return `OtherEGLDisplayAlreadyBound` if called for the same
+    /// `Display` multiple times, as only one context may be bound at any given time.
+    fn unbind_wl_display(&self, display: &Display) -> Result<()>;
```

Comment:
Shouldn't be a `fn(self)` method on `EGLDisplay` rather?

If I understand correctly, calling this method will invalidate a previously created `EGLDisplay`, what happens if I try to continue to use it?

Reply:
Changed it into `EGLDisplay`s Drop implementation